### ISSUE TITLE
Download from Open Subtitles using their REST API

### ISSUE THREAD

**Issue by low-batt** [iina/iina#3920]
**What you want IINA to do:**
Download from Open Subtitles using their new [REST API](https://opensubtitles.stoplight.io/docs/opensubtitles-api/e3750fd63a100-getting-started).

**What IINA does currently:**
Downloads from Open Subtitles using their original [XMLRPC API](https://trac.opensubtitles.org/projects/opensubtitles/wiki/XMLRPC).

**Why you think this should be added:**
Because of this post by Open Subtitles: [OpenSubtitles.org API deprecation and end of life](https://forum.opensubtitles.org/viewtopic.php?f=11&t=17930&sid=d6897753acd72f04f9cf733afe736831) which says:
> After 16 years of service, the time has come to turn off our original API service, to focus our time and energy on the new REST API.

And:
> 2023-12-31 end of life, API endpoint will be turned off

IINA has no choice but to switch to the new REST API before that deadlline.

**Examples of other projects that have something similar:**
N/A

**Comment by low-batt** [comment]
## USERS MUST MIGRATE ACCOUNTS

Open Subtitles [opensubtitles.org](https://www.opensubtitles.org/) is being rebuilt as [opensubtitles.com](https://www.opensubtitles.com/). From the [opensubtitles.com FAQ](https://www.opensubtitles.com/en/faq) question `What is this site ? is it related to opensubtitles.org ?`:
> Yes it definitely is ! opensubtitles.com is the rebuild of opensubtitles.org for the years to come.
>
> Responsive design to fit all screens, UTF-8 conversion and full parsing of all subtitles, AI translations, dynamic framerate, delay or format conversion, and a lot of new features to come.

Thus the mandatory conversion from the Open Subtitles [XMLRPC API](https://trac.opensubtitles.org/projects/opensubtitles/wiki/XMLRPC) to the [REST API](https://opensubtitles.stoplight.io/docs/opensubtitles-api/e3750fd63a100-getting-started) is **not** merely a "behind the scenes" technology change. This is a part of the migration from  [opensubtitles.org](https://www.opensubtitles.org/) (XMLRPC) to [opensubtitles.com](https://www.opensubtitles.com/) (REST API). Open Subtitles is unable to automatically migrate user accounts from to the new [opensubtitles.com](https://www.opensubtitles.com/) site. 

**Users must manually migrate their Open Subtitles account.**

From the [opensubtitles.com FAQ](https://www.opensubtitles.com/en/faq) question `How do I import my previous opensubtitles.org account ?`:
 > Go to the [user import](https://www.opensubtitles.com/users/import) page and fill in the email registered on opensubtitles.org
>
> You will receive an email inviting you set yourself a new password. Your previous uploads will be referenced, user right such as VIP will be imported
>
> **Process is not completely failproof, some cases need to be handled manually. We will be glat to help if you [drop us a note](https://www.opensubtitles.com/contact).**

As a part of this process Open Subtitles will require the account password to be changed. Once finished importing your account you will evenutally need to configure IINA with your new password. At the moment you only need to do that if you update your account on [opensubtitles.org](https://www.opensubtitles.org/) with the new password you configured for the account on [opensubtitles.com](https://www.opensubtitles.com/).

Foillow these instructions to update IINA:
- Start IINA
- Click on  `Settings…` under the `IINA` menu
- The settings panel appears
- On the left side of the panel click on `Subtitle` tab
- On the right side of the panel scroll down to the `Online Subtitles` section
- Click on the `Logout` button
- Log back in specifying your new password

**Comment by low-batt** [comment]
The changes have been merged into the IINA `develop` branch. GitHub [automatically closed](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) the linked issue in reaction to the merge. I am reopening this issue until this change is available in an official release of IINA.

**Comment by low-batt** [comment]
IINA 1.3.4 uses [opensubtitles.com](https://www.opensubtitles.com/) instead of  [opensubtitles.org](https://www.opensubtitles.org/).

**Comment by tsegreto** [comment]
When I use the new API though IINA the results it returns are pretty far off from what I was getting with the old API. For example The.Sopranos.S06E14.Stage.5.720p.WEB-DL.h.264.DD5.1-NTb.mkv will return Sopranos 1x01.

**Comment by opensubtitles** [comment]
I checked that Sopranos, seems good to me:

```
            "file_name": "The.Sopranos.S06E14.HDTV.XviD-NoTV"
            "file_name": "The Sopranos - S06E14 - Stage 5"
            "file_name": "The.Sopranos.S06E14.1080p.Bluray.10Bit.x265.AAC.51-HazMatt"
            "file_name": "The.Sopranos.S06E14.HDTV.XviD-NoTV"
            "file_name": "the.sopranos.s06e14.hr.hdtv.xvid-sitv"
            "file_name": "The Sopranos - S06E14"
            "file_name": "The.Sopranos.S06E14.DVDRip.XviD-WiDE.English"
            "file_name": "The.Sopranos.S06E14.720p.HDDVD.nHD.x264-NhaNc3-eng"
            "file_name": "The.Sopranos.S06E14.1080p.5.1Ch.BluRay.ReEnc-DeeJayAhmed"
            "file_name": "The.Sopranos.S06E14.Stage.5.1080p.HMAX.WEB-DL.DD.5.1.H.264-SLiGNOME"
            "file_name": "The.Sopranos.S06E14.Stage.5.1080p.WEB-DL.h.264.DD5.1-NTb-eng"
            "file_name": "The.Sopranos.S06E14.DVDRip.XviD-TOPAZ"
            "file_name": "The.Sopranos.S06E14.HDTV.XviD-NoTV"
            "file_name": "The.Sopranos.S06E14.720p.HDDVD.x264-BLiND-dut"
            "file_name": "the.sopranos.s06e14.hdtv.xvid-notv"
            "file_name": "The.Sopranos.S06E14.HDTV.XviD-NoTV"
            "file_name": "halcyon-sopranos.s06e14.720p-x264"
            "file_name": "The.Sopranos.S06E14.1080p.5.1Ch.BluRay.ReEnc-DeeJayAhmed-ara"
            "file_name": "The Sopranos - 14 - Stage 5"
            "file_name": "The.Sopranos.S06E14.Stage.5.720p.HDDVD.x264-DiGG"
            "file_name": "[720pMkv.Com]_The.Sopranos.S06E14.Members.Only.480p.BluRay.x264-GAnGSteR-spa"
            "file_name": "The Sopranos - 6x14 - Stage 5"
            "file_name": "The.Sopranos.S06E14.HDTV.XviD-NoTV"
            "file_name": "The.Sopranos.S06E14"
            "file_name": "The.Sopranos.S06E14.HDTV.XviD-NoTV"
            "file_name": "The.Sopranos.S06E14.DVDRip.XviD-WiDE"
            "file_name": "The.Sopranos.S06E14"
            "file_name": "The.Sopranos.S06E14.Stage.5.DVDRip.XviD-WiDE"
            "file_name": "The.Sopranos.S06E14.720p.hddvd.x264.chs"
            "file_name": "The.Sopranos.S06E14.HDTV.XviD-NoTV"
            "file_name": "the.sopranos.s06e14.hdtv.xvid-notv.heb"
            "file_name": "the.sopranos.s06e14.it.subsfactory"
            "file_name": "the.sopranos.s06e14.hdtv.xvid-notv"
            "file_name": "The.Sopranos.S06E14.iNTERNAL.BDRip.x264-ARCHiViST"
            "file_name": "The Sopranos S6 E14 - Stage 5"
            "file_name": "The.Sopranos.S06E14.720p.BluRay.DD5.1.x264-ITSat"
            "file_name": "The.Sopranos.S06E14"
            "file_name": "The.Sopranos.S06E14.DVDRip.XviD-WiDE"
            "file_name": "The.Sopranos.S06E14.BluRay.Subs.German"
            "file_name": "The.Sopranos.S06E14.720p.HDDVD.nHD.x264-NhaNc3"
            "file_name": "The.Sopranos.S06E14.Stage.5.1080p.HMAX.WEB-DL.DD.5.1.H.264-SLiGNOME-NHI"
            "file_name": "The Sopranos s06e14 SWESUB DVDRip XviD - Virus"
            "file_name": "The_Sopranos_6x14"
            "file_name": "The.Sopranos.S06E14.Stage.5.720p.HDDVD.x264-BLiND"
            "file_name": "The.Sopranos.S06E14.DVDRip.XviD-WiDE"
            "file_name": "The.Sopranos.S06E14.HDTV.XviD-NoTV"
            "file_name": "Sopranos S06E14 - Stage 5"
            "file_name": "The.Sopranos.S06E14.DVDRip.XviD-WiDE"
            "file_name": "The.Sopranos.S06E14.DVDRip.XviD-WiDE"
            "file_name": "The.Sopranos.S06E14.DVDRip.XviD-WiDE"
```

anyway, I checked also IINA - would be great to have option for manual string search for movies/tv series

really like REST implementation, devs, contact us, so we can feature article on blog.opensubtitles.com

**Comment by low-batt** [comment]
@opensubtitles Thanks for commenting!

I believe the issue is with the IINA client. I had not noticed the [Guessit](https://opensubtitles.stoplight.io/docs/opensubtitles-api/7783e082edcf7-guessit) API under utilities. Is the intent that the client use this API to extract information such as the season number and episode number to then pass that information as query parameters in the [Search for subtitles
](https://opensubtitles.stoplight.io/docs/opensubtitles-api/a172317bd5ccc-search-for-subtitles) API?

I am currently testing changes do that. For the case at hand that correctly narrows the search results to subtitles for the specific episode. This is being discussed in the new open issue #4749.

IINA 1.3.4 was rushed out to meet the deadline. We are currently working on the fixes that were planed for 1.3.4 and didn't get done in time. The plan is to finish those and make another release soon. This problem will be addressed in that release.

I'm a junior IINA developer. I will let the senior developers know you would like them to contact you.

On the lack of the ability to manually specify a search string, that is only one of several enhancements that should be made to this IINA feature. The plan was to replace the current code with a plugin using IINA's new plugin feature. But the timing did not work out. The plugin system is close to being released, but was not ready in time for this release. So the new code that uses the REST API is a temporary stopgap to meet the deadline. For that reason I did not add any enhancements to the IINA user interface.

**Comment by opensubtitlesdev** [comment]
there was a temporary bug on our guessit system, it didn't detect the season & episodes properly for a few weeks. 

it's all fixed now, you can send your queries like "the sopranos s02e04" it'll be exactly the same as passing the season and episoode numbers separately

**Comment by low-batt** [comment]
@opensubtitles I confirmed searching for subtitles when playing "The.Sopranos.S06E14.Stage.5.720p.WEB-DL.h.264.DD5.1-NTb.mkv" using IINA 1.3.4 now returns results specific to that episode. I will backout my fix that was going into the next release that calls [Guessit](https://opensubtitles.stoplight.io/docs/opensubtitles-api/7783e082edcf7-guessit).

I was wondering how this issue got past me as I was sure I had tested searching for a TV episode and confirmed it was working. I must have done that before the regression on the server occurred.
